libs/dkg/src/pre_dkg_sync.cpp

```diff
@@ -0,0 +1,137 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "dkg/pre_dkg_sync.hpp"
+
+char const *LOGGING_NAME = "Pre-Dkg Sync";
+
+namespace fetch {
+namespace dkg {
+
+PreDkgSync::PreDkgSync(Muddle &muddle, uint8_t channel)
+  : muddle_{muddle}
+  , rbc_{muddle_.AsEndpoint(), muddle_.identity().identifier(), cabinet_,
+         [this](MuddleAddress const &address, ConstByteArray const &payload) -> void {
+           std::set<MuddleAddress>               connections;
+           fetch::serializers::MsgPackSerializer serializer{payload};
+           serializer >> connections;
+           OnRbcMessage(address, connections);
+         },
+         channel}
+{}
+
+void PreDkgSync::OnRbcMessage(MuddleAddress const &from, std::set<MuddleAddress> const &connections)
+{
+  FETCH_LOG_INFO(LOGGING_NAME, "Received connections message");
+  if (peers_.find(from) == peers_.end())
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Message from ", from, " discard as not in peers");
+    return;
+  }
+  if (CheckConnections(connections))
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    if (joined_.find(from) == joined_.end())
+    {
+      joined_.insert(from);
+      ++joined_counter_;
+      lock.unlock();
+      ReceivedDkgReady();
+    }
+    else
+    {
+      FETCH_LOG_WARN(LOGGING_NAME, "Message from ", from, " discard as duplicate message");
+    }
+  }
+  else
+  {
+    FETCH_LOG_WARN(LOGGING_NAME, "Message from ", from, " discard as different connections");
+  }
+}
+
+bool PreDkgSync::CheckConnections(std::set<MuddleAddress> const &connections)
+{
+  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
either this lock doesn't need to be here protecting peers_, or peers_ in onrbcmessage is unprotected